sstable/src/block_reader.rs

```diff
@@ -14,6 +16,29 @@ fn read_u32(read: &mut dyn io::Read) -> io::Result<u32> {
     Ok(u32::from_le_bytes(buf))
 }
 
+struct ReadLimiter<T> {
+    inner: T,
+    budget_left: usize,
+}
+
+impl<T> ReadLimiter<T> {
+    fn new(read: T, budget: usize) -> ReadLimiter<T> {
+        ReadLimiter {
+            inner: read,
+            budget_left: budget,
+        }
+    }
+}
+
+impl<T: Read> Read for ReadLimiter<T> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
```

Comment:
Can we have an implem of `read_exact` here too?

Reply:
(this is mostly defensive)

Reply:
I don't think `read_exact` is ever called on this, the `read_exact` we do is on the `ZstdDecoder`, which leverage a `BufReader`, itself using only `Read::read_buf` (that's all implementation detail).